### DISCUSSION TITLE
Shape library for inserting different shapes quickly

### DISCUSSION THREAD

**Discussion by MiltonBalaOfficial** [xournalpp/xournalpp#6008] [Show and tell]
Hi, I have developed a plugin that enables to insert different 2d, 3d- Shapes, Axix-system, graphs etc.

I am expecting your feedback and comment. Also expecting helps (as I have very little knowledge about coding) to make it more user friendly.

**Download link**
https://github.com/MiltonBalaOfficial/Quick-shape-library

Screenshot:
![image](https://github.com/user-attachments/assets/9069c4bb-64bf-413c-8e62-a4c5c77f29cc)

**Comment by rolandlo** [comment]
Thanks for the nice plugin. Some comments after skimming through the code:
- I would recommend to put the plugin into the user's [config folder](https://xournalpp.github.io/guide/file-locations/), i.e. `C:\Users\<user>\AppData\Local\xournalpp\plugins\` on Windows, since that one can be written without admin privileges. 
- paths like `folderpathGeometric_2D` should be relative to the path of the plugin, which you can get via 
```lua
  local sep = package.config:sub(1, 1) -- path separator depends on OS
  sourcePath = debug.getinfo(1).source:match("@?(.*" .. sep .. ")")
``` 
as in the MigrateFontSizes plugin
- there is a lot of code duplication in `quick_shape_library.lua`. It would be better to create a dictionary for all shapes like 
```lua
local shapes_dict = {
  [1] = {
   name =  "Geometric Two Dimensional Shapes",
   path = "Geometric_2D",
    shapes = {
       [1] = { 
         name = "Arrowhead", 
         path = "ArrowHead.lua",
      }, 
      [2] = { 
       name = "Triangle",
      path = "ScaleneTriangle.lua",
     },
      [2] = { 
       name = "Equilateral Triangle",
      path = "EquilateralTriangle.lua",
     },   
     [4] = {
      name =  "Right Angle Triangle",
      path =  "RightAngleTriangle.lua",
  },  
  [2] = {
    name = "Geometric Three Dimensional Shapes",
    path = "Geometric_3D",
    shapes = { -- etc. }
  }
  -- etc.
}
```
and use it to get rid of all ifs and different callbacks. In a second step this table could be generated automatically from the contents of the Shapes folder.
- `axix` should be `axis`, right?

**Reply by MiltonBalaOfficial** [reply]
Thanks for your guidance. I will try to implement your suggestion ( If my coding knowledge allow me ). I must need your guidance in this regard and please don't get irritated if I ask  simple question about coding. Thanks again

Yes, it should be `axis`

**Reply by MiltonBalaOfficial** [reply]
Hi, I have tried as you suggested but my coding skill stops me midway! I have added the relative path, make the shape dictionary and managed to make the secondary dialog but unable to extract strokes shapeName from the Dictionary. Need help The new file: [https://github.com/MiltonBalaOfficial/Quick-shape-library](https://github.com/MiltonBalaOfficial/Quick-shape-library)

**Reply by rolandlo** [reply]
Use
```lua
function shapeDialogCallback(result)
  local numShapesInCategory = #shapes_dict[currentCategory].shapes
  if result == numShapesInCategory + 1 then
    showMainShapeDialog() -- When "Back" option is selected the secondary dialog closes and main dialog reopens
  elseif result >= 1 and result <= numShapesInCategory then
    local shapeName = shapes_dict[currentCategory].shapes[result].shapeName
    insert_stroke(shapeName)
  end
end
```
where `currentCategory` is a variable initialized outside the functions (like `shapes_dict`) and set at the beginning of the function `showShapeDialog` via 
```lua
currentCategory = categoryIndex
```
Then it all works (except that some physics diagrams are not yet defined).

**Reply by MiltonBalaOfficial** [reply]
Thanks for your guidance and help. I have also added extra codes for extract strokes so that anybody be able to add their own Shapes.
Expecting a final examine and comment form you, thanks

**Reply by rolandlo** [reply]
Thanks for the update. In case you open a PR for including the plugin in Xournal++, I would suggest a lot of more polishing of the code, but I think the plugin is quite useful already. I would advise you to merge the 'read me'.txt into README.md (and use Markdown format), so people read it immediately once they check out your repository. Moreover the paths in the 'read me.txt' are Windows specific. Since there are lots of Linux users among Xournal++ users it would definitely be wise to indicate Linux paths as well. I would put the icons into `~/.local/share/icons/hicolor/scalable/actions/` on Linux.

**Reply by MiltonBalaOfficial** [reply]
I have merged the 'read me'.txt file intoI  README.md using Markdown format, Update the link for icons, but I don't know the Linux path for placing the plugin. I would love to include the plugin in Xournal++, but I don't know anything about pull request. It would be great if you take steps for that purpose as I am unable to do so.

**Reply by rolandlo** [reply]
> I have merged the 'read me'.txt file intoI README.md using Markdown format, Update the link for icons, but I don't know the Linux path for placing the plugin. I would love to include the plugin in Xournal++, but I don't know anything about pull request. It would be great if you take steps for that purpose as I am unable to do so.

The Linux path for placing the plugin is `~/.config/xournalpp/plugins`. Ok, I can create a pull request, keep your name in the plugin.ini, and then add some polish to the code.

**Reply by MiltonBalaOfficial** [reply]
Yes, please add  polish to the code and create a pull request. Actually, without your help it was impossible for me to develop this plugin, most of the credits goes to you. Even I made a plugin for mirroring horizontally and vertically with the code you provided.

**Reply by rolandlo** [reply]
See https://github.com/xournalpp/xournalpp/pull/6020 for a pull request. It's still work in progress.

**Reply by MiltonBalaOfficial** [reply]
Thanks a lot.

**Reply by MiltonBalaOfficial** [reply]
I have detected an issue but don't know the solution. Shouldn't the strokes for a shape be added  serially one by one? For some 3D shapes some strokes need to be added first so that they be at behind the other strokes. Look into this [Test File](https://github.com/MiltonBalaOfficial/Quick-shape-library/blob/main/Test.lua), neither the first stroke nor the last one is added first. Are the strokes added randomly? Is there any solution to add the strokes serial wise so that the 3D shapes can be added with its proper looks.

**Reply by rolandlo** [reply]
The order is unspecified in the current code, but that's easy to fix by using numeric keys for the strokes and `ipairs` instead of `pairs`. See https://stackoverflow.com/questions/55108794/what-is-the-difference-between-pairs-and-ipairs-in-lua for an explanation.

**Reply by MiltonBalaOfficial** [reply]
Yes, got it. If I want to use `ipairs` instead of `pairs` then the `stroke1` , `stroke2`, `stroke3`..... should be changed into `[1]` , `[2]`, `[3]`..... in the shape files and the code should be changed for `function store_stroke_info_in_file()` to do so, right? Some Shapes may need some adjustment for better presentation.

**Reply by rolandlo** [reply]
Yes, I  have already pushed a commit (2b25c8e3eac81cdd8bd38c3ec4fb980da3b887e3) for specifying the order. It would be great if you could suggest adjustments in the order of strokes where needed.

**Reply by MiltonBalaOfficial** [reply]
Here are the modified shapes along with some new [Update1](https://github.com/MiltonBalaOfficial/Quick-shape-library/tree/main/Update%201)

**Comment by Ainson** [comment]
Serendipity this plugin as just what I was looking for, but get an error message on launch:

![Clipboard_11-11-2024_02](https://github.com/user-attachments/assets/9ef4658e-6fda-4fe6-b93c-d22795350bc7)

Maybe it has something to do with paths as I have a portable Xournal installation, away from the default paths described in the instructions, and which the plugin maybe is dependent upon? (I tried to create those paths and put the files there, but no luck)

I did however put the 2 files in the share\icons\hicolor\scalable\actions\ like is said (the icons are visible in Xournal), and the folder with everything else in \Xournal++\share\xournalpp\plugins\ShapeLibrary\. Only confusion was that there was no ShapeLibrary folder as instructed, just the downloaded Quick-shape-library-main, so I renamed it to fit.

Any ideas? Sorry if I missed something obvious.

**Reply by rolandlo** [reply]
Looks like you have the stable version installed and not the latest [nightly build](https://github.com/xournalpp/xournalpp/releases/tag/nightly). On the stable version some API (like `app.openDialog`) is not yet available.

**Comment by Ainson** [comment]
Aha, yes I do have the stable version so that's why then.

The nightly build won't run at all but that's another matter (visual basic runtimes error).
Thx

**Reply by MiltonBalaOfficial** [reply]
I have developed a basic one for the users who use stable version of `Xournal++` . This [Plugin](https://github.com/MiltonBalaOfficial/ZQuickShapes) might be helpful for you.

**Comment by Ainson** [comment]
Works! Thank you.

I was also able add shapes no prob. if I make some, where will I share them exactly?

**Reply by MiltonBalaOfficial** [reply]
A pull request was made for the nightly version #6020, if you want to share your shape there [rolando](https://github.com/rolandlo) might guide you. And if you want your shapes be part of these - [this](https://github.com/MiltonBalaOfficial/ShapeLibrary-for-stable-version-of-Xournal-) and [this](https://github.com/MiltonBalaOfficial/ShapeLibrary) , you might make a pull request or upload your shapes anywhere and just give the link, I will add those. I think [rolando](https://github.com/rolandlo) can provide more technical answer to your question

**Reply by Ainson** [reply]
Ok cheers

**Comment by isaacalbert** [comment]
Great plugin! Thats what i need for school.
Thanks.

**Reply by MiltonBalaOfficial** [reply]
Don't hesitate to raise an issue if there is any problem and if you feel the need of any extra shapes/categories.

**Comment by ebits21** [comment]
Just wanted to chime in that this is a great plugin idea!

I've long wanted the ability to 'stamp' pre-defined shapes and text as I use Xournal to complete hearing test result charts at work. This is very repetitive and I have to write things like 'Cerumen check' many times a day.

It would be great if shapes would add themselves to the config. Also if there was a way to enter the shape name when being made (I see Rolando merged lgi on nightly, maybe this could be used?).

Maybe an option to add shapes to their own layer would be nice.

I'll toy around with it and let you know if I have any issues!

**Reply by rolandlo** [reply]
> It would be great shapes would add themselves to the config. 

This is my goal as well to have shapes being added all from inside Xournal++ (with a few clicks).

>Also if there was a way to enter the shape name when being made (I see Rolando merged lgi on nightly, maybe this could be used?).

The idea I currently have about this is to create a dialog with `lgi` that contains a stack whose pages correspond to the different categories. Each stack page contains a flow box of buttons of fixed size, one for each shape. Each button contains a simplified cairo drawing of the stroke data of the respective shape (or alternatively an image of the shape stored somewhere on the user's device). The dialog would also contain a text entry to enter the name of the shape and a button for inserting the shape. 

It all takes time, but the groundwork from the API side is done, I think. Also we have `lua-lgi` working out of the box in the flatpak, the snap package, the Windows and MacOS version now, so we can really make use of it.

**Reply by ebits21** [reply]
Sounds great!

**Reply by MiltonBalaOfficial** [reply]
> The idea I currently have about this is to create a dialog with `lgi` that contains a stack whose pages correspond to the different categories. Each stack page contains a flow box of buttons of fixed size, one for each shape. Each button contains a simplified cairo drawing of the stroke data of the respective shape (or alternatively an image of the shape stored somewhere on the user's device). The dialog would also contain a text entry to enter the name of the shape and a button for inserting the shape.

It would be amazing if you could provide a prototype with basic code! That way, I can build upon it, even though my coding skills aren’t advanced enough to create a fully upgraded version. With your guidance, I’ll do my best to contribute and bring this cool plugin to life!

**Reply by rolandlo** [reply]
I have pushed a prototype onto the PR #6020. Currently it looks like this:

![Bildschirmfoto vom 2024-12-29 11-39-49](https://github.com/user-attachments/assets/9bd97abb-7a28-4054-b7cf-4fa96c638ec2)

Inserting shapes works, adding hasn't been implemented. The drawing is very basic. I would be glad if you could take it from there. Ask me any questions you have.

Here are some useful resources:
- [GTK 3 docs](https://docs.gtk.org/gtk3/)
- [Lua lgi examples](https://github.com/lgi-devs/lgi/tree/master/samples/gtk-demo)
- [Cairo drawing docs](https://www.cairographics.org/documentation/)

**Reply by rolandlo** [reply]
One thing I'd like to point out:
When the plugin is integrated as a PR, it will be put into a system folder (with no write access for users). So additional shapes  should be put in the folder you get from `app.getFolder("data")` and the config file should be written to (and read from, when it exists) `app.getFolder("config")`

**Reply by MiltonBalaOfficial** [reply]
I will do my best, but it may take me some time to contribute effectively. In the meantime, I encourage you to proceed with upgrading the plugin, as everyone is eager to see it completed as soon as possible. For now, I can assist by providing additional shapes and suggesting modifications to the existing ones. From time to time, I will share my progress on the plugin and reach out for help as needed. I truly appreciate your efforts in advancing the plugin to a more mature and refined version.

**Comment by MiltonBalaOfficial** [comment]
I have developed a better version using `lua-lgi` for my use, I am sharing it for your help to enhance the code and add more options. Again thanks [rolandlo](https://github.com/rolandlo) for `lua-lgi`

**Reply by MiltonBalaOfficial** [reply]
Here it is: [ShapeLibraryWithLgi](https://github.com/MiltonBalaOfficial/ShapeLibraryWithLgi), give your feedback, and if anyone wants to upgrade it, please do as I am not so good at coding to take the plugin to a mature level.

**Reply by rolandlo** [reply]
Thanks for sharing. One remark: For systems where Gtk 4 is installed, you should use
```lua
    Gtk = lgi.require("Gtk", "3.0")
    Gdk = lgi.require("Gdk", "3.0")
```
in order not to load the latest versions.

**Reply by MiltonBalaOfficial** [reply]
Ok, thanks

**Comment by MiltonBalaOfficial** [comment]
I am going towards this direction: 
![image](https://github.com/user-attachments/assets/1efffb9c-fe62-4849-ae63-0755f1b0ff1a)
Though my codes are little bit messy, I managed to make what I desired for a long time. Thanks for your help.

**Comment by MiltonBalaOfficial** [comment]
> The drawing is very basic. 

Drawing quality improved, Scaling also fixed for smaller shapes:
![image](https://github.com/user-attachments/assets/bf615bf1-c678-45e2-bfc2-948bbdb82371)

**Comment by MiltonBalaOfficial** [comment]
[@ rolando](https://github.com/rolandlo)
>When the plugin is integrated as a PR, it will be put into a system folder (with no write access for users). 

What should be the strategy if user wants to update an existing shape (provided with Xournal) ?

**Reply by rolandlo** [reply]
Suggestion: When the user decides to add a shape or to update an existing shape (the first time) the old shapes get copied from the system folder into the user folder` and the preinstalled shapes are ignored. So either config and shapes are both taken from the system folder or both from the user folder, no mixing.

**Reply by atticus-sullivan** [reply]
But wouldn't we make it by this impossible to add shapes via a xournalpp update? Because after the user e.g. added a shape, the shapes are in the user folder and changes made by a xournalpp update won't have any effect anymore.

So wouldn't be "mixing" / merging both sources a better solution? (this way only the explicit shapes created by the user shadows the system/global ones in contrast to shadowing all global shapes)

**Reply by rolandlo** [reply]
> But wouldn't we make it by this impossible to add shapes via a xournalpp update? Because after the user e.g. added a shape, the shapes are in the user folder and changes made by a xournalpp update won't have any effect anymore.
> 
> So wouldn't be "mixing" / merging both sources a better solution? (this way only the explicit shapes created by the user shadows the system/global ones in contrast to shadowing all global shapes)

Hmm, I see your point. Maybe it's better then to add 
- a button for importing (multiple) selected system shapes into the user shapes library
- a checkbox for hiding all system shapes

So a user who is happy with the system shapes library can just add some custom shapes (and not hide system shapes) while a user who isn't happy with some of the system shapes can copy to the user library those he/she likes and add some new ones (while hiding the system shapes most of the time).

What do you think about that solution?

**Reply by atticus-sullivan** [reply]
Sounds like a good idea.

Of course all this increases the complexity of the plugin a bit (it's worth it in my opinion though)

**Reply by MiltonBalaOfficial** [reply]
From a user’s perspective, my initial thought was to create a **dummy plugin** that guides the user through a few setup tasks. The workflow would look something like this:

1. **Open the system folder** where the plugin resides (with functionality to open it programmatically, if possible).  
2. **Copy the necessary files** from this folder.  
3. **Paste the files** into a folder where the user has write access.  

#### Seamless Usage After Initial Setup

Once the user completes this initial setup, the plugin will:

- Detect if the required files have been copied to the desired location.  
- If the setup is complete, the dummy plugin will delete itself and run the actual plugin seamlessly.

#### Why This Approach May Be Suitable

Although users would need to manually copy and paste files during the first setup, subsequent uses would be completely hassle-free. This makes the experience more seamless in the long run.  

I’m sharing this idea as it might inspire a newer, improved solution. While I’m unsure how experienced developers would perceive this approach, it felt like a practical starting point to me.

**Reply by rolandlo** [reply]
That sounds overcomplicated to me. The plugin can just copy the files to the appropriate locations itself (using `app.getFolder("config")` and `app.getFolder("data")` to get the folders.

**Reply by MiltonBalaOfficial** [reply]
> The plugin can just copy the files to the appropriate locations itself (using app.getFolder("config") and app.getFolder("data") to get the folders.

got it!

**Comment by MiltonBalaOfficial** [comment]
# [ShapeLibraryWithLgi](https://github.com/MiltonBalaOfficial/ShapeLibraryWithLgi) - Upgraded Version

This version of the plugin is fully functional and upgraded, offering a comprehensive set of features that meet the needs of most users. While the code may not be fully optimized to professional standards, it includes all the essential functionalities required for shape insertion and customization within Xournal++.

A more refined and solid version is currently under development by **Roland Lötscher** and will be released as part of the official Xournal++ update. In the meantime, please continue to use this version and provide feedback, as your input is valuable for further improvements. Thank you for your help!

![image](https://github.com/user-attachments/assets/a05352b1-5a94-45e0-b87a-9808bd4856ec)
![image](https://github.com/user-attachments/assets/8bdf4cc5-9af7-4ed1-a3c1-21fa47d11985)

* Some new shapes have been added, @rolandlo, I will be glad if you add the new ones as well as the updated ones to the official version.

**Reply by rolandlo** [reply]
Looks pretty nice! Thanks for the work. Just give me some time to finish the official version. I am quite busy right now.

**Comment by searumblinguminari** [comment]
I have a question.
An error message appears and it won't start. I heard that lua lgi is pre-installed in the nightly build, but is there anything else I need to do?
ver. 1.2.8+dev
git commit a3485523
![Error1](https://github.com/user-attachments/assets/f9e70dca-ceb7-4239-a594-9e2836918d04)

**Reply by rolandlo** [reply]
In the nightly build `LuaGObject` (which is a fork of `lgi`) is installed. You can replace "lgi" with "LuaGObject" in the require statement in the Lua script.

**Reply by searumblinguminari** [reply]
Thank you. The problem is solved.
I'm not very familiar with programming or syntax, but after some trial and error I managed to get it to start.
I'm a math teacher and I often draw shapes like parallelograms and pentagons. This plugin is very useful.

**Reply by searumblinguminari** [reply]
I have a question.
I'm talking about a Scalene Triangle, which is included in the Geometric-2D (Category), but I think it's an Isosceles Triangle.
When I rotate it, both sides appear to be the same length.
I'm not an English speaker, so I apologize if I'm wrong.

This plugin allows me to register frequently used table borders, etc., and I find it very convenient. I'm always grateful.

**Reply by searumblinguminari** [reply]
Hello.
I previously asked a question about the error message for ShapeLibraryWithLgi, and you answered it, which resolved the issue and allowed me to use it.
However, after installing the latest nightly build, the same error message reappeared, and it stopped working. Upon checking, I found that the function plotter also stopped working. I don't know when the function plotter stopped working.
Were there any changes in this update that might be related to this issue?
I apologize for asking, but could you please let me know?

xournalpp-1.3.4+dev-nightly-20260503-windows-setup-AMD64.exe
(7163bc9)
<img width="653" height="253" alt="error1" src="https://github.com/user-attachments/assets/8f12c9ec-d99e-4d0b-8b57-f6c79e112982" />
<img width="659" height="251" alt="error2" src="https://github.com/user-attachments/assets/f55ffc4a-25a8-4ed3-9cca-65f226a5b948" />

**Reply by rolandlo** [reply]
Thanks for reporting. Yes, there were some changes with the Lua version used (5.4->5.5). I will check when arriving home.

**Reply by rolandlo** [reply]
The issue should be fixed in the next nightly build.

**Reply by searumblinguminari** [reply]
It worked perfectly. Thank you.
